src/ai/backend/manager/api/session.py

```diff
@@ -508,52 +509,70 @@ async def _create(request: web.Request, params: Any) -> web.Response:
-                            kernels.c.service_ports,
                         ])
                         .select_from(kernels)
                         .where(kernels.c.id == kernel_id)
                     )
                     result = await conn.execute(query)
-                    row = result.first()
-                if row['status'] == KernelStatus.RUNNING:
-                    resp['status'] = 'RUNNING'
-                    for item in row['service_ports']:
-                        response_dict = {
-                            'name': item['name'],
-                            'protocol': item['protocol'],
-                            'ports': item['container_ports'],
-                        }
-                        if 'url_template' in item.keys():
-                            response_dict['url_template'] = item['url_template']
-                        if 'allowed_arguments' in item.keys():
-                            response_dict['allowed_arguments'] = item['allowed_arguments']
-                        if 'allowed_envs' in item.keys():
-                            response_dict['allowed_envs'] = item['allowed_envs']
-                        resp['servicePorts'].append(response_dict)
-                else:
-                    resp['status'] = row['status'].name
+
+                return result.first()
+
+            async def _update_progress(
+                app: web.Application,
+                source: AgentId,
+                event: KernelPullProgressEvent,
+            ) -> None:
+                # update both current and total
+                progress[0] = int(event.current_progress)
+                progress[1] = int(event.total_progress)
+
+            root_ctx.event_dispatcher.subscribe(KernelPullProgressEvent, request.app, _update_progress)
+            kernel_id = resp['sessionId']
+            while True:
+                result = await _get_status(kernel_id)
+                if result is None:
+                    continue
+                if result['status'] == KernelStatus.PREPARING:
+                    await reporter.update(0)
+                if result['status'] == KernelStatus.RUNNING:
+                    break
+                reporter.current_progress = progress[0]
+                reporter.total_progress = progress[1]
+                await reporter.update(0)
+            await asyncio.sleep(0.5)
+
+        task_id = await root_ctx.background_task_manager.start(
+                kernelpullprogress,
+                name='kernel-pull-progress'
+        )
+        resp['background_task'] = str(task_id)
+        resp['sessionName'] = str(params['session_name'])
+        resp['status'] = 'PENDING'
+        resp['servicePorts'] = []
+        resp['created'] = True
+
+        app_ctx.pending_waits.add(current_task)
+        max_wait = params['max_wait_seconds']
+        try:
+            if max_wait > 0:
+                with timeout(max_wait):
+                    await asyncio.sleep(0.5)
```

Comment:
This breaks the existing API semantics.

Reply:
In the existing API semantics, after enqueue-session is finished, manager sends the response to client. But in this situation, since everything related to create-session is completed and client obtain the response from manager, client does not know bgtask_id during pulling kernel. So, I think that client requests twice, one of them is creating session, the other is getting the information of the existing session.

Reply:
Hmmm. I don't think we need to add another column to the kernels table.

Let's do it as follows:
* When the client calls the session creation API with `enqueue_only` option, the bgtask ID should be included in the response.
  - It may or may not finish soon, but it is the client's job to keep track of session status afterwards.
* When the client calls the session creation API with `max_wait` option:
  - If the creation finishes within the max-wait timeout, just return the full result including service port information and etc. as before. It's usually the case when the image is already populated on the assigned agent(s), and in this case you don't have to reply back the bgtask ID.
  - If not, then include the bgtask ID in the response so that the client could keep track of the image pulling progress.

Reply:
![image](https://user-images.githubusercontent.com/555156/132480630-38878edd-0166-48d0-b69e-fdd79d841de7.png)

Reply:
Oh, sorry.. I remove the code to add column of progress on table. I use Only KernelPullProgressEvent! 
I'll try to implement your feedback.

Reply:
I have a question.
- Your comment 'I think we may need to keep track of the bgtask ID in the kernels table so that ...' means that we add a column of 'bgtask-id' to 'kernels' table?

Reply:
Ah, sorry for confusion. That's a twist in my head... 🙄  I don't think we need to add bgtask ID to the kernels table.